lib/icinga/checkable-dependency.cpp

```diff
@@ -7,22 +7,181 @@
 
 using namespace icinga;
 
-void Checkable::AddDependency(const Dependency::Ptr& dep)
+/**
+ * Register all the dependency groups of the current Checkable to the global dependency group registry.
+ *
+ * Initially, each Checkable object tracks locally its own dependency groups on Icinga 2 startup, and once the start
+ * signal of that Checkable is emitted, it pushes all the local tracked dependency groups to the global registry.
+ * Once the global registry is populated with all the local dependency groups, this Checkable may not necessarily
+ * contain the exact same dependency groups as it did before, as identical groups are merged together in the registry,
+ * but it's guaranteed to have the same *number* of dependency groups as before.
+ *
+ * Note: Re-registering the very same dependency groups to global registry will crash the process with superior error
+ * messages that aren't easy to debug, so make sure to never call this method more than once for the same Checkable.
+ */
+void Checkable::PushDependencyGroupsToRegistry()
 {
-	std::unique_lock<std::mutex> lock(m_DependencyMutex);
-	m_Dependencies.insert(dep);
+	std::lock_guard lock(m_DependencyMutex);
+	decltype(m_DependencyGroups) dependencyGroups{0, HashDependencyGroup, CompareDependencyGroups};
+	m_DependencyGroups.swap(dependencyGroups);
+
+	for (auto& dependencyGroup : dependencyGroups) {
+		m_DependencyGroups.emplace(DependencyGroup::Register(dependencyGroup));
+	}
 }
 
-void Checkable::RemoveDependency(const Dependency::Ptr& dep)
+std::vector<DependencyGroup::Ptr> Checkable::GetDependencyGroups() const
 {
-	std::unique_lock<std::mutex> lock(m_DependencyMutex);
-	m_Dependencies.erase(dep);
+	std::unique_lock lock(m_DependencyMutex);
+	return {m_DependencyGroups.begin(), m_DependencyGroups.end()};
+}
+
+/**
+ * Add the provided dependency to the current Checkable list of dependencies.
+ *
+ * The "modifiedGroup" parameter is used as an output variable to store the dependency group that has been modified.
+ * Meaning, either the dependency groups identity has changed after adding the new dependency to it, or the provided
+ * dependency caused the creation of a new dependency group and decoupling from of the existing one. In the latter case,
+ * the output variable will be set to the previously existing group that the current Checkable was member of.
+ *
+ * @param dependency The dependency to add.
+ * @param refreshGlobalRegistry Whether to automatically refresh the global registry or not.
+ * @param modifiedGroup The dependency group that has been modified.
+ */
+void Checkable::AddDependency(const Dependency::Ptr& dependency, bool refreshGlobalRegistry, DependencyGroup::Ptr* modifiedGroup)
```

Comment:
You're not using the return value of that function so far, so I'd just do that instead of an output parameter.